### PR TITLE
parser:  support multi bracket for subquery

### LINKED ISSUE #28476: TiDB v5.2.1 Parser Unexpected Error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
mysql> SELECT COUNT(*) FROM plan_executions WHERE (EXISTS((SELECT * FROM triggers WHERE plan_executions.trigger_id=triggers.id AND triggers.type='CRON')));
```

### 2. What did you expect to see? (Required)
If I change the "(())" after 'EXISTS' into "()", there will be no error: 
```sql
mysql> SELECT COUNT(*) FROM plan_executions WHERE (EXISTS(SELECT * FROM triggers WHERE plan_executions.trigger_id=triggers.id AND triggers.type='CRON'));
+----------+
| COUNT(*) |
+----------+
|        3 |
+----------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)

An error: 
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 146 near "))"

### 4. What is your TiDB version? (Required)

V5.2.1 (In previous version of tidb--v5.1.1, It is OK to execute )

### LINKED ISSUE #29416: Set operation UNION with parentheses to order precedence is not supported in TiDB
## Bug Report

Please answer these questions before submitting your issue. Thanks!
```sql
drop table if exists t1,t2,t3,t4;
create table t1 (a int not null, b char (10) not null);
insert into t1 values(1,‘a’),(2,‘b’),(3,‘c’),(3,‘c’);
select * from (((select * from t1)) union (select * from t1) union (select * from t1)) a;
```
### 2. What did you expect to see? (Required)
```sql
mysql> drop table if exists t1,t2,t3,t4;
Query OK, 0 rows affected, 2 warnings (0.05 sec)

mysql> create table t1 (a int not null, b char (10) not null);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t1 values(1,'a'),(2,'b'),(3,'c'),(3,'c');
Query OK, 4 rows affected (0.00 sec)
Records: 4  Duplicates: 0  Warnings: 0

mysql> select * from (((select * from t1)) union (select * from t1) union (select * from t1)) a;
+---+---+
| a | b |
+---+---+
| 1 | a |
| 2 | b |
| 3 | c |
+---+---+
3 rows in set (0.00 sec)
```
### 3. What did you see instead (Required)
```sql
mysql> drop table if exists t1,t2,t3,t4;
Query OK, 0 rows affected, 2 warnings (0.42 sec)

mysql> create table t1 (a int not null, b char (10) not null);
Query OK, 0 rows affected (0.08 sec)

mysql> insert into t1 values(1,'a'),(2,'b'),(3,'c'),(3,'c');
Query OK, 4 rows affected (0.02 sec)
Records: 4  Duplicates: 0  Warnings: 0

mysql> select * from (((select * from t1)) union (select * from t1) union (select * from t1)) a;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 41 near "union (select * from t1) union (select * from t1)) a" 
```
### 4. What is your TiDB version? (Required)
```sql
+-------------------------+--------------------------------------------------------------------------+
| Variable_name           | Value                                                                    |
+-------------------------+--------------------------------------------------------------------------+
| innodb_version          | 5.6.25                                                                   |
| protocol_version        | 10                                                                       |
| tidb_analyze_version    | 2                                                                        |
| tidb_row_format_version | 2                                                                        |
| tls_version             | TLSv1,TLSv1.1,TLSv1.2                                                    |
| version                 | 5.7.25-TiDB-v5.2.1                                                       |
| version_comment         | TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible |
| version_compile_machine | x86_64                                                                   |
| version_compile_os      | osx10.8                                                                  |
+-------------------------+--------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by sylzd**
### What problem does this PR solve?

Issue Number: close #29416 
also close #28476
Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by sylzd** [comment]
@lzmhhh123 could you give some suggestions? since I thinks it's simillar to https://github.com/pingcap/parser/issues/1127

**Comment by lzmhhh123** [comment]
For the `shift/reduce` error, I have no idea either.

**Comment by bb7133** [comment]
This PR also close https://github.com/pingcap/tidb/issues/28476, please add the case from it:

```
SELECT COUNT(*) FROM plan_executions WHERE (EXISTS((SELECT * FROM triggers WHERE plan_executions.trigger_id=triggers.id AND triggers.type='CRON')));
```
and
```
select exists((select 1));
```

File: parser/parser.y [hunk 1]
```diff
@@ -9151,6 +9151,29 @@ SubSelect:
 		rs.SetText(src[yyS[yypt-1].offset:yyS[yypt].offset])
 		$$ = &ast.SubqueryExpr{Query: rs}
 	}
+|	'(' SubSelect ')'
+	{
+		subQuery := $2.(*ast.SubqueryExpr).Query
+		isRecursive := true
+		for isRecursive {
+			if _, isRecursive = subQuery.(*ast.SubqueryExpr); isRecursive {
```
Comment (bb7133) [inline parser/parser.y:9159]: Please add a comment here: this is a trick to remove redundant, as what you've written in the test case(https://github.com/pingcap/tidb/pull/29656/files#diff-6eaacb6a6fa7a8b6851fa3bc9917a3cfc6a7a3e191069b6d59ae5385b9beb6faR4344)

**Comment by bb7133** [comment]
LGTM

**Comment by bb7133** [comment]
PTAL @lzmhhh123 @kennytm @tangenta

File: parser/parser.y (hunk 1 above)
Comment (sylzd) [reply parser/parser.y:9160]: ok~

**Review by tangenta — APPROVED** [review]

**Review by bb7133 — APPROVED** [review]

**Comment by bb7133** [comment]
This PR close #28053 also.

**Pull request merged**